xchange-stream-dydx/src/main/java/info/bitrich/xchangestream/dydx/dto/v3/dydxInitialOrderBookMessage.java

```diff
@@ -0,0 +1,91 @@
+package info.bitrich.xchangestream.dydx.dto.v3;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import info.bitrich.xchangestream.dydx.dto.dydxWebSocketTransaction;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.knowm.xchange.dto.trade.LimitOrder;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.SortedMap;
+
+import static info.bitrich.xchangestream.dydx.dydxStreamingAdapters.dydxOrderBookChanges;
+
+/** Author: Max Gao (gaamox@tutanota.com) Created: 08-03-2021 */
+@NoArgsConstructor
+@Getter
+@Setter
+public class dydxInitialOrderBookMessage extends dydxWebSocketTransaction {
+  @JsonProperty("contents")
+  private Contents contents;
+
+  @Getter
+  @Setter
+  public static class Contents {
+    @JsonProperty("offset")
+    private String offset;
+
+    @JsonProperty("bids")
+    private Order bids[];
+
+    @JsonProperty("asks")
+    private Order asks[];
+  }
+
+  @Getter
+  @Setter
+  public static class Order {
+    @JsonProperty("price")
+    private String price;
+
+    @JsonProperty("size")
+    private String size;
+  }
+
+  public OrderBook toOrderBook(
+      SortedMap<BigDecimal, BigDecimal> bids,
+      SortedMap<BigDecimal, BigDecimal> asks,
+      int maxDepth,
+      CurrencyPair currencyPair) {
+    String[][] bidsData = null;
+    String[][] asksData = null;
+
+    if (this.contents.getBids() != null) {
+      bidsData =
+          Arrays.stream(this.contents.getBids())
+              .map(b -> new String[] {b.price, b.size})
+              .toArray(String[][]::new);
```

Comment:
One thing that is sticking out to me here is the fact that in the initial orderbook image you don't deal with the id maps https://github.com/knowm/XChange/pull/3975/files#diff-f6a6e1dea42f3fbe31c7b3849de8fc9e532281de9769e4cfc153d463322bbd53R68, doesn't that mean that after the initial snapshot you won't have anything in that map even though you will have layers in the orderbook? Doesn't seem right but maybe something related to the api.

Reply:
V3 orderbook messages follow the more intuitive orderbook update structure that most other exchanges have. For some reason, V1 uses an id in place of prices when returning updates. 

https://legacy-docs.dydx.exchange/?json#orderbook

This is an example of non-snapshot V1 messages:
```{
  "type": "channel_data",
  "connection_id": "0b88ebf2-98ec-4de6-b781-62e1648ae657",
  "message_id": 1278,
  "channel": "orderbook",
  "id": "WETH-DAI",
  "contents": {
    "updates": [
      {
        "type": "REMOVED",
        "id": "0xe8aad0c4ea79e014b398da781a1b711a4c12086348cd421e4f5c19ba077c3374",
        "side": "BUY"
      },
      {
        "type": "UPDATED",
        "id": "0x510f0a948e9b4488bfcc3b5aa2abf28334e174825a56fa2c6948fd5aaf3dd420",
        "side": "SELL",
        "amount": "50600000000000000000"
      }
    ]
  }
}
```

vs V3 non-snapshot messages:

```
{
  "type": "channel_data",
  "id": "BTC-USD",
  "connection_id": "e2a6c717-6f77-4c1c-ac22-72ce2b7ed77d",
  "channel": "v3_orderbook",
  "message_id": 2,
  "contents": {
    "offset": "178",
    "bids": [["102", "12"]],
    "asks": [["104", "0" ]]
  }
}
```